### PR TITLE
Enable cross-compilation from Linux to Mac

### PR DESCRIPTION
**Pull request by mherrmann**
The following commands can be used to test the new implementation:

```
git clone git@github.com:brave/brave-browser.git
cd brave-browser
git clone git@github.com:brave/brave-core.git src/brave

... check out a revision of b-c with the changes from this PR...

npm install
npm config set target_os=mac
npm config set target_arch=x64
npm run init
npm run build -- Static
```

Resolves https://github.com/brave/devops/issues/7551. Depends on https://github.com/brave/Sparkle/pull/19/ and https://github.com/brave/devops/pull/7615.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

This change could affect Brave's auto-update functionality on macOS. Please test the scenarios 1.-4. specified under "Test Plan" in https://github.com/brave/brave-core/pull/12335. It should be enough to do this on one or two macOS versions, say the most recent one and the oldest one supported by Brave.

### CODE REVIEW

**Comment by simonhong** [comment]
With this change, sparkle will be built separated from b-c?

**Comment by mherrmann** [comment]
@simonhong yes. See https://github.com/brave/Sparkle/pull/19.

**Review by simonhong — APPROVED** [review]
LGTM for sparkle related changes.

**Comment by mherrmann** [comment]
We would like to merge this. Even if it doesn't end up getting used in CI, it can help developers (like me) whose primary OS is Linux. @goodov @bridiver can I ask for your review?

File: build/commands/lib/config.js
Comment (goodov) [inline build/commands/lib/config.js:364]: I don't think `ios` can be built that way at all, so we shouldn't support it here.

File: script/download_rust_deps.py
Comment (mherrmann) [reply script/download_rust_deps.py]: Fixed in 6a8a169590d40a6d292802db9610f4cca0c00c9c.

File: build/mac/download_hermetic_xcode.py
Comment (mherrmann) [reply build/mac/download_hermetic_xcode.py]: Fixed in 0c5ef4eefadfc51ad461f720ba819dfd791884ad.

File: build/commands/lib/config.js
Comment (mherrmann) [reply build/commands/lib/config.js]: Fixed in d610a52518070148ce5e65708fe065f65d4df3b8.

File: patches/chrome-tools-build-mac-BUILD.gn.patch
Comment (mherrmann) [reply patches/chrome-tools-build-mac-BUILD.gn.patch:11]: Your suggested changes work for me. However: What's the advantage of making it a single-line patch? The patch "region" is already contiguous, even tough it's multiple lines. So there shouldn't be any more conflicts than if it were on a single line, no?

Comment (goodov) [reply patches/chrome-tools-build-mac-BUILD.gn.patch:11]: We always try to do single-line patches, it's easier to maintain. @bridiver can share some details why we prefer it this way.

Comment (mherrmann) [reply patches/chrome-tools-build-mac-BUILD.gn.patch:11]: Thank you. I found "ignore lint line length rules to squash patches onto one line whenever possible" in the Wiki [here](https://github.com/brave/brave-browser/wiki/Patching-Chromium#patch-the-chromium-files). But it doesn't give an explanation as to why. (Maybe it's because then the line numbers in the patch file stay relatively constant?) Would love to hear about the motivation for single-line patches from @bridiver, which I'd then add to the Wiki. In the meantime, I'll just do it this way.

Comment (mherrmann) [reply patches/chrome-tools-build-mac-BUILD.gn.patch]: Resolved in 17d0938f0a8d447bf6ff61fa1201edcdeff78115.

**Comment by goodov** [comment]
can you pls build a release builds on this branch (x64 and arm64) to make sure everything's working fine?

**Comment by mherrmann** [comment]
@goodov sure. There you go:

 * [x64](https://ci.brave.com/job/test-brave-browser-build-macos-x64/651/)
 * [arm64](https://ci.brave.com/job/test-brave-browser-build-macos-arm64/179/)

**Comment by goodov** [comment]
> @goodov sure. There you go:
> 
> * [x64](https://ci.brave.com/job/test-brave-browser-build-macos-x64/651/)
> * [arm64](https://ci.brave.com/job/test-brave-browser-build-macos-arm64/179/)

Thanks. Looks good. Sorry for being too picky, but Release builds don't use Goma. Can you please check that these jobs also work with `USE_GOMA` disabled?

**Comment by mherrmann** [comment]
@goodov no problem. Here are the builds without GOMA:

- https://ci.brave.com/job/test-brave-browser-build-macos-x64/652/parameters/
- https://ci.brave.com/job/test-brave-browser-build-macos-arm64/180/parameters/

**Review by goodov — APPROVED** [review]

File: patches/chrome-tools-build-mac-BUILD.gn.patch
Comment (bridiver) [inline patches/chrome-tools-build-mac-BUILD.gn.patch:9]: Why does the toolchain matter for constants? In any case if we are touching this we should convert to a dep with a public config and put it in "//brave/tools/build/mac:infoplist_strings_util"

**Comment by mherrmann** [comment]
@brave/sec-team we are downloading [two new Rust dependencies](https://github.com/brave/brave-core/pull/13537/files#diff-9b571efa8c7252e8079d783391ba1504f35ae68edad7c4e60f869dc377e782d6R182) as part of Brave's build process. There are already [several others](https://github.com/brave/brave-core/blob/44efc62c58b7eff28b6101445dd89f3c53e8a73d/script/download_rust_deps.py#L186-L201) which we are downloading. Do we need a security review for this?

(This was raised by @bridiver.)

File: patches/chrome-tools-build-mac-BUILD.gn.patch
Comment (mherrmann) [reply patches/chrome-tools-build-mac-BUILD.gn.patch:9]: (As discussed on Slack) Because we're building a tool here, which by default gets compiled with the host toolchain. In this rare case, we need Mac constants on Linux, thus need to depend on the default toolchain.

**Review by bridiver — APPROVED** [review]
approved based on discussion of keeping the dep for sparkle so we retain the option to build locally

**Comment by diracdeltas** [comment]
> @brave/sec-team we are downloading [two new Rust dependencies](https://github.com/brave/brave-core/pull/13537/files#diff-9b571efa8c7252e8079d783391ba1504f35ae68edad7c4e60f869dc377e782d6R182) as part of Brave's build process. There are already [several others](https://github.com/brave/brave-core/blob/44efc62c58b7eff28b6101445dd89f3c53e8a73d/script/download_rust_deps.py#L186-L201) which we are downloading. Do we need a security review for this?

yes please, this should be quick though. https://github.com/brave/security/issues

**Comment by mherrmann** [comment]
I created the security review issue: https://github.com/brave/security/issues/924

**Review by kdenhartog — APPROVED** [review]
security issue was approved by @diracdeltas so going to approve here on behalf of the sec-team. LGTM

**Review by kdenhartog — APPROVED** [review]
re-approving incase it's needed for merge. I can't remember exactly how stale review dismissals work

File: build/mac/download_sparkle.py
Comment (mihaiplesa) [inline build/mac/download_sparkle.py:19]: `sparkle/sparkle-1.0.tar.gz`

**Review by mihaiplesa — CHANGES_REQUESTED** [review]

**Comment by mherrmann** [comment]
This is again ready to be merged from my POV after another round of reviews. Recent changes are the introduction of a `--build_sparkle` build parameter to `npm`. This is default false and decides whether the Sparkle macOS update framework should be built from source or whether pre-built binaries should be used instead. The logic for building Sparkle is thus now again contained in brave-core, as requested by @bridiver.

I think I will merge this PR on Wednesday, unless there are any objections.

**Review by mihaiplesa — APPROVED** [review]

File: build/commands/lib/config.js
Comment (bridiver) [inline build/commands/lib/config.js:185]: Defaults should be set in gn, not config.js and `build_sparkle` can be set with `--gn`

File: build/commands/lib/config.js
Comment (bridiver) [inline build/commands/lib/config.js:403]: I believe this can be simplified to just `this.targetArch === 'x64' && this.targetOS === 'linux'` since we're already excluding android

Comment (bridiver) [reply build/commands/lib/config.js:403]: also if we're cross compiling macos on linux, I believe this would incorrectly set use_vaapi to true

Comment (bridiver) [reply build/commands/lib/config.js:403]: actually this whole vaapi thing looks suspect to me and I think we need to revisit it

Comment (bridiver) [reply build/commands/lib/config.js:403]: also pretty sure this should be `use_vaapi_x11`, not `use_vaapi`

```
# VA-API should also be compiled by default on x11-using linux devices
  # using x86/x64.
  use_vaapi = use_vaapi_x11 || (is_chromeos_lacros &&
                                (target_cpu == "x86" || target_cpu == "x64"))
```

File: build/commands/lib/config.js
Comment (mherrmann) [reply build/commands/lib/config.js]: Resolved in cddafe61ee2bd06c379bd19a5346181274f02094.

File: build/commands/lib/config.js
Comment (mherrmann) [reply build/commands/lib/config.js]: I fixed the `if` statement as you suggested in f2a3c0b1ac727477d22ac1984469af4c26111089.

> if we're cross compiling macos on linux, I believe this would incorrectly set use_vaapi to true

I actually don't think the implementation you reviewed would have done this. The reason is that `targetOS` is set to `mac` when we cross-compile.

> also pretty sure this should be use_vaapi_x11, not use_vaapi

It's possible. I don't yet understand why we need to set `use_vaapi` or `use_vaapi_x11` at all, given that they have dynamic defaults that appear reasonable.

```
  # Indicates if X11 VA-API-based hardware acceleration is to be used.
  # See also the comment near the |use_vaapi| arg.
  use_vaapi_x11 =
      is_linux && ozone_platform_x11 && !is_chromecast && !is_chromeos_lacros &&
      (target_cpu == "x86" || target_cpu == "x64")
```

The approach I'd choose is to create a follow-up issue to investigate this.

**Review by bridiver — APPROVED** [review]

**Comment by mherrmann** [comment]
Just a quick update that I would merge this but there have been CI problems with Windows and I want a green build before merging. I expect to get it and merge after my PTO next week.

**Pull request merged**
